Source/MSBuild.Community.Tasks/DeleteTree.cs

```diff
@@ -192,38 +291,28 @@ internal static IList<string> MatchDirectories(string pattern)
                             pathRegex.Append(singleCharRegex);
                     }
                     partStart = wildIndex + 1;
-                }       
+                }
             }
             pathRegex.Append("$");
 
-            Regex searchRegex = new Regex(pathRegex.ToString(), RegexOptions.IgnoreCase);
-            string[] dirs = Directory.GetDirectories(rootPath, "*", SearchOption.AllDirectories);
-            foreach (string dir in dirs)
-                if (searchRegex.IsMatch(dir))
-                    directories.Add(dir);
+            var searchRegex = new Regex(pathRegex.ToString(), RegexOptions.IgnoreCase);
+            var dirs = Directory.GetDirectories(rootPath, "*", SearchOption.AllDirectories);
+
+            directories.AddRange(dirs.Where(dir => searchRegex.IsMatch(dir)));
 
             return directories;
         }
 
         private static string FindRootPath(string[] parts, out int pathIndex)
         {
-            string rootPath;
-            List<string> root = new List<string>();
-            foreach (string part in parts)
-            {
-                if (part.Equals(recursiveDirectoryMatch))
-                    break;
-
-                if (part.IndexOfAny(wildcardCharacters) >= 0)
-                    break;
-
-                root.Add(part);
-            }
-
-            if (root.Count == 0)
-                rootPath = Environment.CurrentDirectory;
-            else
-                rootPath = string.Join(Path.DirectorySeparatorChar.ToString(), root.ToArray());
+            var root = parts
```

Comment:
The new implementation is not equivalent to the old. The correct version of the new implementation should be:

       var root = parts
          .TakeWhile(part => !part.Equals(recursiveDirectoryMatch) && part.IndexOfAny(wildcardCharacters) < 0)
          .ToList();